### ISSUE TITLE
unstable computation of standard-deviation

### ISSUE THREAD

**Issue by alan-isaac** [NetLogo/NetLogo#1723]
This is most easily illustrated by example.  Put the following in the `Code` tab:

```
patches-own [attr]
to setup
  ask patches [set attr one-of [0 1]]
end
```

Add the following monitor to the interface:

`standard-deviation [attr] of patches`

After `setup`, you will see the monitor values fluctuate: the order of patches matters for the value of the computation.  This should either be documented or a numerically stable algorithm should be adopted.  (Of course, a simple fix is to use `standard-deviation map [? -> {attr] of ?] sort patches], but explaining this to undergrads is not fun.)

**Comment by SethTisue** [comment]
Agentsets are always processed in random order. I think this is well documented already. It's something every NetLogo programmer needs to learn about agentsets pretty much as soon as they are introduced.

In contexts where we want to focus students' attention on the content of a model rather than how floating point math works, I would recommend setting the precision of the monitor so not all decimal places are displayed, and the flickering will go away.

In other contexts, where students are learning programming, IEEE 754 is just a fact of life, it crops up as soon as you do `0.1 + 0.1 + 0.1`, even.

In short I think this is normal, expected behavior.

**Comment by alan-isaac** [comment]
Of course that's the source of the problem, as my illustrated fix demonstrates.  But providing a more computationally stable algorithm for the variance (& thus standard-deviation) would not require any other changes in NetLogo.  This might even more generally be seen as a good thing (TM).  However, if not, documenting the dependence of the computation on the order of the list items seems reasonable.  So if you insist that the behavior is not buggy (despite the availability of fixes), then the algorithm used and this implication might be documented.  I can attest that students find it surprising (since it cam up in class today).  Cheers.

**Comment by SethTisue** [comment]
fwiw, the current implementations are here:
* https://github.com/NetLogo/NetLogo/blob/hexy/netlogo-headless/src/main/prim/etc/_standarddeviation.java
* https://github.com/NetLogo/NetLogo/blob/hexy/netlogo-headless/src/main/prim/etc/_variance.java

(I'm not sure whether to link to the netlogo-headless or netlogo-gui versions, I'm surprised that multiple versions even exist.)

For lack of the relevant math knowledge, I don't know what order-independent versions might look like.

**Comment by SethTisue** [comment]
> the algorithm used and this implication might be documented

I don't understand the point of doing this given that just about anything involving floating point numbers is likely to come out differently in the last decimal place or two, depending.

**Comment by alan-isaac** [comment]
The problem is that this two-pass algorithm is known to be numerically unstable, so this is really a different case.  For true order independence, you would have to sort.  (This is probably affordable and a good idea in most NetLogo models.)  But there are cheapter alternatives.  Reasonable would be to report only 15 digits of precision, since there won't generally be more than that (I believe).  A more standard response (I believe) is to use compensated summation: https://en.wikipedia.org/wiki/Kahan_summation_algorithm

**Comment by SethTisue** [comment]
Perhaps a pull request substituting a stable algorithm would be accepted. The only reason I can think of it might _not_ be accepted is if the performance were not just a little worse, but a lot worse. (Where the line is or ought to be, I don't know.)
